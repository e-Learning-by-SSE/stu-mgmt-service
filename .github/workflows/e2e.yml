name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
