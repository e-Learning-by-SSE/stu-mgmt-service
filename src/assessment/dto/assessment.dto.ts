import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
