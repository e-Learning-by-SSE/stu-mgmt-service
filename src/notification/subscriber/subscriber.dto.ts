import { ApiProperty } from "@nestjs/swagger";
import { Event } from "../../course/events/events";

export type SubscribedEvents = { [key in Event]?: boolean } & { ALL?: boolean };

export class SubscriberDto {
	@ApiProperty({ description: "Name of your application." })
	name: string;

	@ApiProperty({
		description:
			"Complete URL to an endpoint in your application that receives HTTP-POST messages."
	})
	url: string;

	@ApiProperty({
		type: Object,
		description:
			"Map of events that you want to be notified about. Refer to the Event-Enum for event names. You can use 'ALL' to subscribe to all events.",
		example: `{"${Event.USER_JOINED_GROUP}": true,
		"${Event.ASSIGNMENT_STATE_CHANGED}": true
		}`
	})
	events: SubscribedEvents;

	@ApiProperty({ description: "Date of the creation or latest update to the subscription." })
	updateDate?: Date;
}
