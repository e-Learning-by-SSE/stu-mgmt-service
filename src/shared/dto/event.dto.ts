export class EventDto {
	event: string;
	timestamp: Date;
	payload?: Record<string, unknown>;
}
