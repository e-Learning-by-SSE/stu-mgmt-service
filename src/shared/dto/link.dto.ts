export class LinkDto {
	url: string;
	name: string;
}
