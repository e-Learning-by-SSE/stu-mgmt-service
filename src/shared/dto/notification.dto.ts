import { AssignmentId } from "../../course/entities/assignment.entity";
import { CourseId } from "../../course/entities/course.entity";
import { GroupId } from "../../course/entities/group.entity";
import { Event } from "../../course/events/events";
import { UserId } from "../entities/user.entity";

export class NotificationDto {
	event: Event;
	courseId: CourseId;
	assignmentId?: AssignmentId;
	groupId?: GroupId;
	userId?: UserId;
	payload?: Record<string, unknown>;
}
