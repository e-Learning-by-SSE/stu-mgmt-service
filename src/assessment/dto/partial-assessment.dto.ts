import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
