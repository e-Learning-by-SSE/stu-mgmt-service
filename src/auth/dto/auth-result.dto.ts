import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
