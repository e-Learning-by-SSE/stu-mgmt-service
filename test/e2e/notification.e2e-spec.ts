import { SubscriberDto } from "../../src/notification/subscriber/subscriber.dto";
import { COURSE_INFO_2_2020 } from "../mocks/courses.mock";
import { SUBSCRIBER_1_INFO_SOSE2020, SUBSCRIBER_MOCK } from "../mocks/subscribers.mock";
import { TestSetup } from "../utils/e2e";

const course = COURSE_INFO_2_2020;
const baseUrl = "/notifications";

describe("Notification E2E", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	describe("PUT-REQUESTS of NotificationController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createUsers();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createSubscribers();
		});

		describe("(PUT) /notifications/courses/{courseId}/subscribers/{name}", () => {
			const route = (courseId: string, name: string) =>
				`${baseUrl}/courses/${courseId}/subscribers/${name}`;

			it("Not subscribed -> Adds subscriber", () => {
				const subscriber: SubscriberDto = {
					name: "mySubscriber",
					events: { ALL: true },
					url: "http://example.url"
				};

				return setup
					.request()
					.put(route(course.id, subscriber.name))
					.send(subscriber)
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubscriberDto;
						expect(result.name).toEqual(subscriber.name);
						expect(result.events).toEqual(subscriber.events);
						expect(result.url).toEqual(subscriber.url);
					});
			});

			it("Already subscribed -> Updates subscriber", () => {
				const subscriber = SUBSCRIBER_1_INFO_SOSE2020;

				return setup
					.request()
					.put(route(course.id, subscriber.name))
					.send(subscriber)
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubscriberDto;
						expect(result.name).toEqual(subscriber.name);
						expect(result.events).toEqual(subscriber.events);
						expect(result.url).toEqual(subscriber.url);
					});
			});
		});
	});

	describe("GET-Requests of NotificationController (e2e)", () => {
		beforeAll(async () => {
			await setup.clearDb();
			await setup.dbMockService.createUsers();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createSubscribers();
		});

		describe("(GET) /notifications/courses/{courseId}/subscribers", () => {
			const route = (courseId: string) => `${baseUrl}/courses/${courseId}/subscribers`;

			it("Retrieves all subscribers", () => {
				const expected = SUBSCRIBER_MOCK.map(mock => ({
					...mock.dto,
					updateDate: null
				}));

				return setup
					.request()
					.get(route(course.id))
					.expect(200)
					.expect(({ body }) => {
						const result = body as SubscriberDto[];
						const resultWithoutDate = result.map(res => ({ ...res, updateDate: null }));
						expect(resultWithoutDate).toMatchSnapshot(expected);
					});
			});

			it("No subscribers -> Empty array", () => {
				return setup
					.request()
					.get(route("java-wise1920"))
					.expect(200)
					.expect(({ body }) => {
						expect(body).toEqual([]);
					});
			});
		});
	});

	describe("DELETE-Requests of NotificationController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createUsers();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createSubscribers();
		});

		describe("(DELETE) /notifications/courses/{courseId}/subscribers/{name}", () => {
			const route = (courseId: string, name: string) =>
				`${baseUrl}/courses/${courseId}/subscribers/${name}`;

			it("Subscribed -> Removes subscriber", () => {
				const subscriber = SUBSCRIBER_1_INFO_SOSE2020;

				return setup.request().delete(route(course.id, subscriber.name)).expect(200);
			});

			it("Subscriber does not exists -> Throws 404", () => {
				return setup.request().delete(route(course.id, "does-not-exist")).expect(404);
			});
		});
	});
});
