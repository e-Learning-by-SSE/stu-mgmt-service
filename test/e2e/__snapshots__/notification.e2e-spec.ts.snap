// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification E2E GET-Requests of NotificationController (e2e) (GET) /notifications/courses/{courseId}/subscribers Retrieves all subscribers 1`] = `
Array [
  Object {
    "events": Object {
      "ALL": true,
    },
    "name": "subscriber_1",
    "updateDate": null,
    "url": "http://example-one.url",
  },
  Object {
    "events": Object {
      "GROUP_REGISTERED": true,
      "GROUP_UNREGISTERED": true,
      "USER_JOINED_GROUP": true,
      "USER_LEFT_GROUP": true,
    },
    "name": "subscriber_2",
    "updateDate": null,
    "url": "http://example-two.url",
  },
]
`;
